### PR TITLE
Support arm64/v9 and minor variants

### PR DESCRIPTION
**Pull request by harryzcy**
Related https://github.com/containerd/containerd/issues/10022.

This follows updated "Platform Variants" from the spec: https://github.com/opencontainers/image-spec/pull/1172

### CODE REVIEW

File: compare.go
Comment (tianon) [inline compare.go]: I keep linking to https://github.com/golang/go/issues/60905#issuecomment-1611867608 to describe this, but I imagine (vain hope, perhaps?) there's probably a better / more "canonical" reference for this complexity? :sob:

File: compare.go
Comment (tianon) [inline compare.go]: This `variant[:2]` worries me -- would it be better/safer to do something like an explicit `strings.Cut` on `.`?

File: compare.go
Comment (tianon) [inline compare.go:133]: Shouldn't this latter half be `v9`, not `v8`? (giving further credence to perhaps using `strings.Cut` above instead, for which the right-hand-side would just be the correct answer?)

Perhaps I'm missing what this fallback block is for. (Apologies!)

Comment (harryzcy) [reply compare.go:133]: I left the comment in code:

```go
// All arm64/v8.x and arm64/v9.x are compatible with arm/v8 (32-bits) and below.
// There's no arm64 v9 variant, so it's normalized to v8.
```

File: compare.go
Comment (harryzcy) [reply compare.go]: It's mentioned in https://github.com/golang/go/issues/60905#issuecomment-1611815533:

From [ARM Architecture Reference Manual](https://developer.arm.com/documentation/ddi0487/latest/), section A2.3 describes "Armv9-A architecture extensions", and then its subsections has the following:

- A2.3.1 The Armv9.0 architecture extension
  - It is Armv8.5 compliant.
- A2.3.2 The Armv9.1 architecture extension
  - It is Armv8.6 compliant.
  - It is Armv9.0 compliant.
- A2.3.3 The Armv9.2 architecture extension
  - It is Armv8.7 compliant.
  - It is Armv9.1 compliant.
- A2.3.4 The Armv9.3 architecture extension
  - It is Armv8.8 compliant.
  - It is Armv9.2 compliant.
- A2.3.5 The Armv9.4 architecture extension
  - It is Armv8.9 compliant.
  - It is Armv9.3 compliant.

Looks like there's no Armv9.5 yet

**Comment by harryzcy** [comment]
@tianon Hi, can you help to review this again?

**Comment by AkihiroSuda** [comment]
cc @tonistiigi

File: database.go
Comment (dmcgowan) [inline database.go]: With "9.6" now, will this need to be updated every year since we can't really predict if there will be "9.7"?

File: compare.go
Comment (dmcgowan) [reply compare.go]: Announced by not equivalent 8.10 or 8.11 for 9.6?

File: compare.go
Comment (dmcgowan) [inline compare.go:101]: Should this stop at 8.9?

File: database.go
Comment (harryzcy) [reply database.go]: Fixed in 8771e0d74df0784730c2276e64756d2dfb17fa32. it now allows any v9.x version

File: compare.go
Comment (harryzcy) [reply compare.go:101]: Fixed in fd2a630aedb409576b25be2f23730ddc3e099eb1

File: platforms.go
Comment (dmcgowan) [inline platforms.go]: This is specifying the same character differently than the line below `A-Za-z0-9_.-`. Just a nit, could clean up these regexps later

**Review by dmcgowan — APPROVED** [review]

**Review by AkihiroSuda — DISMISSED** [review]
Please squash the commits

**Review by estesp — APPROVED** [review]
LGTM; although if this gets any more complicated it might be easier to have some form of lookup table rather than trying to follow a set of for loops?

**Review by samuelkarp — APPROVED** [review]
LGTM per discussion in the maintainer channel, though +1 to @estesp's comment; this is probably much easier to maintain if we convert it to a lookup table rather than math and loops.

**Review by dmcgowan — APPROVED** [review]

**Comment by vincentbernat** [comment]
What are the next steps to support `--platform=linux/arm64/v8.2` in both Docker and buildx?

Edit: it seems buildx is happy with `--platform=linux/arm64/v8.2`.
Edit: I am able to push the right variant to GHCR registry: https://github.com/akvorado/akvorado/pkgs/container/akvorado/508205260
Edit: docker pull won't download it unless explicitely asked for

So, I suppose the next step is to add support in Docker:

```go
// maximumSpec returns the distribution platform with maximum compatibility for the current node.
func maximumSpec() ocispec.Platform {
	p := platforms.DefaultSpec()
	if p.Architecture == "amd64" {
		p.Variant = archvariant.AMD64Variant()
	}
	return p
}
```

This looks like a good place for that. The difficulty is to find a library that is able to provide the variant for the current CPU. This does not seem to exist.

**Comment by harryzcy** [comment]
@vincentbernat `platforms.DefaultSpec()` is provided by this repo, which only sets `Variant` to `v8`.

It's quite difficult to analyze the instruction/feature support for each variant. The source of truth is https://developer.arm.com/documentation/109697/2025_06/Feature-descriptions/The-Armv8-0-architecture-extension and so on.

Some other languages like Rust is doing this as well: https://github.com/rust-lang/rust/blob/d1ed52b1f5b78bf66127b670af813b84d57aeedb/src/tools/rust-analyzer/crates/hir-ty/src/target_feature.rs#L143. But that may not necessarily be correct in our context, e.g. `FEAT_RDM` is marked as optional in doc.

If we were to do it ourselves, one way is to use `"golang.org/x/sys/cpu"`. But it doesn't support all the features we want to detect. e.g. `FEAT_FlagM` is mandatory in armv8.4, but the `"golang.org/x/sys/cpu"` does not have it yet.

**Pull request merged**
